src/coreclr/vm/appdomain.cpp

```diff
@@ -3352,20 +3352,37 @@ BOOL AppDomain::AddFileToCache(AssemblySpec* pSpec, PEAssembly * pPEAssembly, BO
     }
     CONTRACTL_END;
 
-    GCX_PREEMP();
-    DomainCacheCrstHolderForGCCoop holder(this);
-
-    // !!! suppress exceptions
-    if(!m_AssemblyCache.StorePEAssembly(pSpec, pPEAssembly) && !fAllowFailure)
     {
-        // TODO: Disabling the below assertion as currently we experience
-        // inconsistency on resolving the Microsoft.Office.Interop.MSProject.dll
-        // This causes below assertion to fire and crashes the VS. This issue
-        // is being tracked with Dev10 Bug 658555. Brought back it when this bug
-        // is fixed.
-        // _ASSERTE(FALSE);
+        GCX_PREEMP();
+        DomainCacheCrstHolderForGCCoop holder(this);
+
+        // !!! suppress exceptions
+        if(!m_AssemblyCache.StorePEAssembly(pSpec, pPEAssembly) && !fAllowFailure)
+        {
+            // TODO: Disabling the below assertion as currently we experience
+            // inconsistency on resolving the Microsoft.Office.Interop.MSProject.dll
+            // This causes below assertion to fire and crashes the VS. This issue
+            // is being tracked with Dev10 Bug 658555. Brought back it when this bug
+            // is fixed.
+            // _ASSERTE(FALSE);
+
+            EEFileLoadException::Throw(pSpec, FUSION_E_CACHEFILE_FAILED, NULL);
+        }
+    }
 
-        EEFileLoadException::Throw(pSpec, FUSION_E_CACHEFILE_FAILED, NULL);
+    DomainAssembly *pParentAssembly = pSpec->GetParentAssembly();
+    BINDER_SPACE::Assembly *pBinderSpaceAssembly = pPEAssembly->GetHostAssembly();
```

Comment:
This should probably check `HasHostAssembly` / that `pBinderSpaceAssembly` is not null. Some cases (dynamically emitted assemblies is one, I think) might not have one - should we be adding the reference in those cases too?.

Reply:
If there is no host assembly, how can we get the `LoaderAllocator` of such an assembly? It seems it is possible to return a dynamically emitted collectible assembly from the ALC.Load / Resolving event handler.

Reply:
You should be able to `GetAssemblyBinder` and then get the `LoaderAllocator` off off that.

Reply:
For the default ALC/binder (`AssemblyBinder::IsDefault=true`), I think the loader allocator on the binder will be null, but we shouldn't need the reference in that case.

Reply:
There is a problem with GetAssemblyBinder usage in general - if the assembly was marked for unload before it was returned from the ALC.Load / event, the binder had the LoaderAllocator already set to NULL. I have to check if it is a problem for the dynamic assemblies too.

Reply:
Ok, so it seems it is not possible to return dynamic assembly from the ALC.Load / resolving event. Attempt to do so results in fail fast in both cases:
```
Unhandled exception. System.IO.FileLoadException: Could not load file or assembly 'TestInterface, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'. An operation is not legal in the current state. (0x80131509 (COR_E_INVALIDOPERATION))
File name: 'TestInterface, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
 ---> System.InvalidOperationException: Dynamically emitted assemblies are unsupported during host-based resolution.
   at System.Reflection.RuntimeAssembly.<GetType>g____PInvoke__|26_0(QCallAssembly assembly, UInt16* name, Int32 throwOnError, Int32 ignoreCase, ObjectHandleOnStack type, ObjectHandleOnStack keepAlive, ObjectHandleOnStack assemblyLoadContext)
   at System.Reflection.RuntimeAssembly.GetType(QCallAssembly assembly, String name, Boolean throwOnError, Boolean ignoreCase, ObjectHandleOnStack type, ObjectHandleOnStack keepAlive, ObjectHandleOnStack assemblyLoadContext)
   at System.Reflection.RuntimeAssembly.GetType(String name, Boolean throwOnError, Boolean ignoreCase)
   at Test.Load2() in C:\Users\janvorli\source\repos\ExtremeUnloadTest\Program.cs:line 139
   at Test.TestFullUnload(Int32 testCase) in C:\Users\janvorli\source\repos\ExtremeUnloadTest\Program.cs:line 194
   at Test.Main(String[] args) in C:\Users\janvorli\source\repos\ExtremeUnloadTest\Program.cs:line 227
```

Reply:
Ohh, right - sorry, I forgot we explicitly block that. There should be no concern here then.

Reply:
It's true right now, but it's another sort of artificial limitation. It's true that non-unloadable ALCs should not return assembly from unloadable ALC in their Load, but unloadable ALC should be able to do so, there's no real problem in that case, we just need to express the dependency between the ALCs.

Reply:
I've just added an assert that the pBinderSpaceAssembly  is not NULL for the future in case we want to change this.

Reply:
I'm going to regret this comment... but I don't get it:
Technically every time we ask to resolve an assembly name to an assembly instance, the question has to be in a context of some ALC. Always. Since the only thing which can actually resolve the name is an ALC. If there's no ALC apparent (like loading CoreLib at the very begining), we default to the Default ALC, but there's always one when we start the bind.
Let's call that ALC the "requesting ALC". When an assembly is resolved it means it has to already be loaded and thus it belongs to an ALC (again each assembly always belongs to an ALC). Let's call that ALC the "resolved ALC".

So what we want to do is "If the resolved ALC is collectible add a reference from requesting ALC to the resolved ALC".

(This also means that we should never add that reference if requesting ALC is non-collectible as that would root the collectible resolved ALC forever - and thus this case should cause a failure - which it does).

I know that in the code we don't have the ALC at hand like this and instead we have the "parent assembly" and "resolved assembly" (which is probably not true for CoreLib, but maybe CoreLIb doesn't go through this code path as it's special). But the logic is the same. So this means that (except for CoreLib) parent assembly should never be NULL.

I guess I'm just repeating what the code does now... but I had to think through it this way.

Anyway - I think this is worth a comment in the code. Not the "create reference" that one is relatively easy to see, but when we expect NULLs and that it should basically never happen.

Reply:
> I know that in the code we don't have the ALC at hand like this and instead we have the "parent assembly" and "resolved assembly"

I think the confusing part here is exactly this translation. As you noted, whenever we resolve an assembly name, it has to be in the context of some ALC. However, how we project that information is rather dependent on how that resolution was requested.

> So this means that (except for CoreLib) parent assembly should never be NULL.

Right now, when it is a load based on an assembly/type reference, the parent assembly will be non-null. But we don't actually set the parent assembly when doing an explicit load or loading via reflection (in some cases) - for example, using `Assembly.Load`, `AssemblyLoadContext.LoadFromAssemblyName`, or `Activator.CreateInstance(<assemblyName>, <typeName>)`. In those cases, I believe we do set the 'fallback binder' (`AssemblyBinder`) on the spec as the projection of the managed ALC in the native runtime.

Reply:
So, do we have anything actionable left on this PR? 
I can't tell for sure form the comments if all concerns were addressed.

Reply:
There is @vitek-karas' request (which I'd second) for documenting expectations:

> I think this is worth a comment in the code. Not the "create reference" that one is relatively easy to see, but when we expect NULLs

Reply:
Does the following look good?

```
There are situations when `pParentAssembly` is NULL. For example when assembly is loaded via reflection. It is ok   
to ignore those.    
The purpose of the following code is to catch scenarios when resolution happens in another ALC solely based on  
ALC.Load or ALC.Resolving event handler. In such cases we need to add a reference here as it is not guaranteed  
to be added otherwise. These are also cases when `pParentAssembly` will always be set.
```

Reply:
I think it would be useful to call out both 'parent assembly' for the request and the 'result assembly' for the result. Thoughts on something like the below?

```
If the ALC for the result (pPEAssembly) is collectible and not the same as the ALC for the request (pSpec),
which can happen when extension points (AssemblyLoadContext.Load, AssemblyLoadContext.Resolving) resolve the
assembly, we need to ensure the result ALC is not collected before the request ALC. We do this by adding an
explicit reference between the LoaderAllocators corresponding to the ALCs.

To get the LoaderAllocators, we rely on the DomainAssembly corresponding to:
- Parent assembly for the request
  - For loads via explicit load or reflection, this will be NULL. In these cases, the request ALC should not
    implicitly (as in not detectable via managed references) depend on the result ALC, so it is fine to not
    add the reference.
  - For loads via assembly references, this will never be NULL.
- Result assembly for the result
  - For dynamic assemblies, there will be no host assembly, so we have the result assembly / LoaderAllocator.
    We currently block resolving to dynamic assemblies, so we simply assert that we have a host assembly.
  - For non-dynamic assemblies, we should be able to get the host assembly.
```

Reply:
This looks better indeed. These details may not be obvious.

Reply:
Thanks!